src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -143,6 +141,72 @@ static bool JoinRestrictionListExistsInContext(JoinRestriction *joinRestrictionI
 											   joinRestrictionContext);
 
 
+/*
+ * QueryContainsDistributionKeyEquality returns true if either
+ *    (i)  there exists join in the query and all relations joined on their
+ *         partition keys
+ *    (ii) there exists only union set operations and all relations has
+ *         partition keys in the same ordinal position in the query
+ */
+bool
+QueryContainsDistributionKeyEquality(PlannerRestrictionContext *plannerRestrictionContext,
+									 Query *originalQuery)
```

Comment:
minor: I find "Contains" a bit confusing (does it contain 1 distribution key equality?).  Maybe `AllDistributionKeysInQueryAreEqual`.

Super minor: it seems a bit more natural/common that the query comes first and the context after.

Reply:
I liked `AllDistributionKeysInQueryAreEqual `